bpf/ut/icmp_port_unreachable_test.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019-2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ut_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/felix/bpf/nat"
+)
+
+func TestICMPPortUnreachable(t *testing.T) {
+	RegisterTestingT(t)
+
+	_, ipv4, _, _, pktBytes, err := testPacketUDPDefault()
+	Expect(err).NotTo(HaveOccurred())
+
+	runBpfUnitTest(t, "icmp_port_unreachable.c", func(bpfrun bpfProgRunFn) {
+		res, err := bpfrun(pktBytes)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(res.Retval).To(Equal(0))
+
+		Expect(res.dataOut).To(HaveLen(134)) // eth + ip + 64 + udp + ip + icmp
+
+		pktR := gopacket.NewPacket(res.dataOut, layers.LayerTypeEthernet, gopacket.Default)
+		fmt.Printf("pktR = %+v\n", pktR)
+
+		checkICMPPortUnreachable(pktR, ipv4)
+	})
+
+}
+
+func TestNATNoBackendFromHEP(t *testing.T) {
+	RegisterTestingT(t)
+
+	iphdr := *ipv4Default
+
+	_, ipv4, l4, _, pktBytes, err := testPacket(nil, &iphdr, nil, nil)
+	Expect(err).NotTo(HaveOccurred())
+
+	udp := l4.(*layers.UDP)
+
+	natkey := nat.NewNATKey(ipv4.DstIP, uint16(udp.DstPort), uint8(ipv4.Protocol)).AsBytes()
+	err = natMap.Update(
+		natkey,
+		nat.NewNATValue(0, 1, 0, 0).AsBytes(),
+	)
+	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		err := natMap.Delete(natkey)
+		Expect(err).NotTo(HaveOccurred())
+	}()
+
+	runBpfTest(t, "calico_from_host_ep", rulesDefaultAllow, func(bpfrun bpfProgRunFn) {
+		res, err := bpfrun(pktBytes)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(res.RetvalStr()).To(Equal("TC_ACT_UNSPEC"), "expected program to return TC_ACT_UNSPEC")
+
+		pktR := gopacket.NewPacket(res.dataOut, layers.LayerTypeEthernet, gopacket.Default)
+		fmt.Printf("pktR = %+v\n", pktR)
+
+		checkICMPPortUnreachable(pktR, ipv4)
+	})
+
+	err = natMap.Update(
+		natkey,
+		nat.NewNATValue(0, 0, 0, 0).AsBytes(),
```

Comment:
nit: perhaps a comment how this case and the case above differ in what they test.